### PR TITLE
Fix multi resolving expressions in composite steps

### PR DESCRIPTION
**Pull request by ElenaKornievich**

### CODE REVIEW

File: jbehave-core/src/main/java/org/jbehave/core/steps/StepCreator.java [hunk 1]
```diff
@@ -958,11 +958,27 @@ private void parametriseStep() {
             stepMatcher.find(stepWithoutStartingWord);
             ParameterName[] names = parameterNames(method);
             Type[] types = parameterTypes(method, names);
-            List<Pair<String, String>> parameterValuesPairsList = parameterValuesForStep(namedParameters, types, names, true);
+            Map<String, String> compositeParameters = new HashMap<>();
+            namedParameters.keySet().forEach(k -> {
+                try
+                {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/StepCreator.java]: `try {`

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 2]
```diff
@@ -33,6 +37,48 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+        compositeCounter.get().incrementAndGet();
+    }
+
+    @Override
+    public void successful(String step) {
+        processCompositeMap();
+    }
+
+    private void processCompositeMap()
+    {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: `private void processCompositeMap()  {`

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 3]
```diff
@@ -33,6 +37,48 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+        Map<String, Object> compositeObjectsMap = compositeObjects.get();
+        if (compositeCounter.get().decrementAndGet() == 0 && !compositeObjectsMap.isEmpty())
+        {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: `if (compositeCounter.get().decrementAndGet() == 0 && !compositeObjectsMap.isEmpty()) {`

File: jbehave-core/src/test/java/org/jbehave/core/steps/CompositeCandidateStepsBehaviour.java [hunk 4]
```diff
@@ -2,8 +2,12 @@
 
 import com.thoughtworks.paranamer.BytecodeReadingParanamer;
 import org.jbehave.core.annotations.*;
+import org.jbehave.core.configuration.Configuration;
+import org.jbehave.core.configuration.Keywords;
```
Comment (valfirst) [inline jbehave-core/src/test/java/org/jbehave/core/steps/CompositeCandidateStepsBehaviour.java]: noisy change

File: jbehave-core/src/test/java/org/jbehave/core/steps/CompositeCandidateStepsBehaviour.java [hunk 5]
```diff
@@ -180,15 +188,26 @@ public void shouldMatchCompositesAndCreateComposedNestedSteps() {
...
         assertThat(composedSteps.size(), equalTo(2));
+        int i = 0;
         for (Step step : composedSteps) {
-            step.perform(mock(StoryReporter.class), null);
+            if(i==0) {
+                stepsContext.putCompositeObject("location", "left");
+            }
+            else {
+                stepsContext.putCompositeObject("location", "top");
+            }
```
Comment (valfirst) [inline jbehave-core/src/test/java/org/jbehave/core/steps/CompositeCandidateStepsBehaviour.java]:
```java
stepsContext.putCompositeObject("location", i==0 ? "left" : "top");
```

File: jbehave-core/src/test/java/org/jbehave/core/steps/CompositeCandidateStepsBehaviour.java [hunk 6]
```diff
@@ -180,15 +188,26 @@ public void shouldMatchCompositesAndCreateComposedNestedSteps() {
...
+            i++;
+            step.perform(mock(StepsContext.class), null);
+
             List<Step> nestedComposedSteps = ((StepCreator.ParametrisedStep) step).getComposedSteps();
+            //stepsContext.putCompositeObject("location", nestedComposedSteps.);
```
Comment (valfirst) [inline jbehave-core/src/test/java/org/jbehave/core/steps/CompositeCandidateStepsBehaviour.java]: useless comment

File: jbehave-core/src/main/java/org/jbehave/core/steps/StepCreator.java [hunk 7]
```diff
@@ -5,11 +5,7 @@
 import java.lang.reflect.Method;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (uarlouski) [inline jbehave-core/src/main/java/org/jbehave/core/steps/StepCreator.java]: useless changes

**Comment by By-ivan** [comment]
Commit message unclear

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 8]
```diff
@@ -33,6 +37,46 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+    public void beforeStep(String step) {
+        compositeCounter.get().incrementAndGet();
+    }
+
+    @Override
+    public void successful(String step) {
+        processCompositeMap();
+    }
+
+    private void processCompositeMap() {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: `processCompositeMap` -> `stopStep`

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 9]
```diff
@@ -33,6 +37,46 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+        Map<String, Object> compositeObjectsMap = compositeObjects.get();
+        if (compositeCounter.get().decrementAndGet() == 0 && !compositeObjectsMap.isEmpty()) {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: `!compositeObjectsMap.isEmpty()` looks redundant, is it really required?

File: jbehave-core/src/main/java/org/jbehave/core/steps/StepCreator.java [hunk 10]
```diff
@@ -958,9 +959,22 @@ private void parametriseStep() {
             stepMatcher.find(stepWithoutStartingWord);
             ParameterName[] names = parameterNames(method);
             Type[] types = parameterTypes(method, names);
+            namedParameters.keySet().forEach(k -> {
+                try {
+                    Optional.ofNullable(stepsContext.getCompositeObject(k.replaceAll("[<>]", "")))
+                            .ifPresent(v -> namedParameters.put(k, (String) v));
+                } catch (StepsContext.ObjectNotStoredException e) {
```
[the remark this answers is not in the document]
Comment (By-ivan) [reply jbehave-core/src/main/java/org/jbehave/core/steps/StepCreator.java]: not fixed ?

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 11]
```diff
@@ -33,6 +37,46 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+        compositeCounter.get().incrementAndGet();
+    }
+
+    @Override
+    public void successful(String step) {
+        stopStep();
+    }
+
+    private void stopStep() {
+        Map<String, Object> compositeObjectsMap = compositeObjects.get();
```
Comment (uarlouski) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: move it inside spring?

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 12]
```diff
@@ -33,6 +37,46 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+    public void beforeStep(String step) {
+        compositeCounter.get().incrementAndGet();
+    }
+
+    @Override
+    public void successful(String step) {
+        stopStep();
+    }
+
+    private void stopStep() {
```
Comment (uarlouski) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: code will look better if we place this method after pending(String)

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 13]
```diff
@@ -33,6 +37,46 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+
+    @Override
+    public void pending(String step) {
+        stopStep();
+    }
+
+    private void stopStep() {
+        if (compositeCounter.get().decrementAndGet() == 0) {
+            Map<String, Object> compositeObjectsMap = compositeObjects.get();
+            compositeObjectsMap.clear();
```
Comment (uarlouski) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: compositeObjects.get()clear

File: jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java [hunk 14]
```diff
@@ -33,6 +37,46 @@ public void put(String key, Object object, ToContext.RetentionLevel retentionLev
...
+    }
+
+    @Override
+    public void pending(String step) {
+        stopStep();
+    }
+
+    private void stopStep() {
+        if (compositeCounter.get().decrementAndGet() == 0) {
+            Map<String, Object> compositeObjectsMap = compositeObjects.get();
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/context/StepsContext.java]: so now it can be inlined

**Pull request merged**